### PR TITLE
X509 Path Validation Flag to Ignore Root Certificate Lifetime

### PR DESCRIPTION
**Pull request by FAlbertDev**
As suggested in issue #1089, this pull request introduces a new feature that allows the path validation process to ignore the lifetime of trusted root certificates. Specifically, the `Path_Validation_Restrictions` now includes an `ignore_root_lifetime` option. When this option is set to `true`, the trusted root certificate can be expired or not yet valid. It's important to note that this option only applies to the trusted root certificate and not to any intermediate or leaf certificates.

### Online Research
I believe that [this answer on Stack Exchange](https://security.stackexchange.com/a/120471) provides a comprehensive explanation of why expired CA certificates may still be allowed. It's worth noting that many cryptographic libraries also permit the use of expired CA certificates, as mentioned in issue #1089.

Additionally, the BSI's technical guideline [BSI TR-02103](https://www.bsi.bund.de/SharedDocs/Downloads/DE/BSI/Publikationen/TechnischeRichtlinien/TR02103/BSI-TR-02103.pdf?__blob=publicationFile&v=2#page=43) in section 8.2.2.4 states the following (translated):

>In path validation according to [RFC 5280], there is no provision for checking the validity period of a trust anchor when using the contents of a trust anchor according to [RFC 5937]. Therefore, the timely removal of a trust anchor that is no longer considered valid must be ensured through other means.

Also [RFC 5280](https://datatracker.ietf.org/doc/html/rfc5280#section-6.1) states in section 6.1:

>When the trust anchor is provided in the form of a self-signed certificate, this self-signed certificate is not included as part of the prospective certification path.

As far as I understand it, it all boils down to what it means for an application to mark a root certificate as trusted. Does it simply imply trust in the public key within the certificate, regardless of its lifetime? Or does it imply trust in the certificate as long as it remains valid? In my opinion, it is best to let the application make this decision.

However, we need to determine the default option. For now, I have chosen the "security by default" approach, where expired root certificates are disallowed by default. It is worth noting that both cases pass all tests in Botan's test suite. Please feel free to share any concerns you may have.

### Test
I have added a simple test case that utilizes a trusted, self-signed root certificate that is valid from 2022 to 2028. Additionally, a derived end-entity certificate that is valid from 2020 to 2030 is included. I have configured both options and verified that the verification process succeeds or fails as expected. It is worth noting that Botan's other test cases also cover scenarios where only the intermediate or leaf certificate is expired.

### Other
It would be greatly appreciated if we could merge this pull request before the 3.4.0 release, as it is required for our application.

### CODE REVIEW

File: src/tests/data/x509/misc/root_cert_time_check/README.md [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+## Test Certificates
+This test case contains two certificates:
+- A trusted root certificate `root.crt`. Validity range (years): 2022-2028.
+- An end-entity certificate `leaf.crt` chaining to `root.crt`.
+  Validity range (years): 2020-2030.
+
+These certificates are used to test Botan's behavior for verification at
+specific time points. For example, verification in 2025 succeeds,
+verification in 2031 fails, and verification at 2029 depends on the option.
```
[the remark this answers is not in the document]
Comment (FAlbertDev) [reply src/tests/data/x509/misc/root_cert_time_check/README.md:17]: Uff, it's quite hacky C++ code. However, I don't think the full code is necessary. I'll try it 👍

**Comment by reneme** [comment]
As @randombit mentioned [here](https://github.com/randombit/botan/issues/1089#issuecomment-370563630), perhaps it makes sense to set a warning-status if the root is outside its validity interval but the end-entity is considered valid.

File: src/tests/data/x509/misc/root_cert_time_check/README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (reneme) [reply src/tests/data/x509/misc/root_cert_time_check/README.md:17]: Doesn't need to be spotless. Chances are that it remains buried in this Readme for all eternity. ;)

[the remark this answers is not in the document]
Comment (reneme) [reply src/tests/data/x509/misc/root_cert_time_check/README.md:17]: Fair.

**Comment by FAlbertDev** [comment]
I added the warning. Thanks for the hint. I'll look into the CI tomorrow :)

File: src/lib/x509/x509path.h [hunk 2]
```diff
@@ -50,13 +50,16 @@ class BOTAN_PUBLIC_API(2, 0) Path_Validation_Restrictions final {
...
+      *        time range of the trusted root certificate only produce warnings
       */
       Path_Validation_Restrictions(
          bool require_rev = false,
          size_t minimum_key_strength = 110,
          bool ocsp_all_intermediates = false,
          std::chrono::seconds max_ocsp_age = std::chrono::seconds::zero(),
-         std::unique_ptr<Certificate_Store> trusted_ocsp_responders = std::make_unique<Certificate_Store_In_Memory>());
+         std::unique_ptr<Certificate_Store> trusted_ocsp_responders = std::make_unique<Certificate_Store_In_Memory>(),
+         bool ignore_trusted_root_time_range = false);
```
Comment (randombit) [inline src/lib/x509/x509path.h:62]: Unrelated to this PR, this constructor is getting out of hand - we might want to consider introducing a builder-style API for this.

**Review by randombit — APPROVED** [review]
Seems fine to me to support this, thanks for the test also

File: src/lib/x509/x509path.h (hunk 2 above)
Comment (reneme) [reply src/lib/x509/x509path.h:62]: I've created a ticket (#3939) to collect API improvement ideas for X.509.

**Comment by randombit** [comment]
Ready to merge?

**Comment by FAlbertDev** [comment]
I did a minor history cleanup. After CI is done, I'll merge.

**Pull request merged**